### PR TITLE
Anilist V2 Import!

### PR DESCRIPTION
**Pull request by toyhammered**
# What
Update Anilist import to integrate with V2 (graphql)

# Why
Anilist has deprecated their V1 api, thus the previous import I wrote doesn't work anymore. I will be following a very similar structure to all our other imports in utilizing the ListImport Model.

# Checklist

- [ ] All files pass Rubocop
- [x] Any complex logic is commented
- [x] Any new systems have thorough documentation
- [x] Any user-facing changes are behind a feature flag (or: explain why they can't be)

- [x] All the tests pass
- [x] Tests have been added to cover the new code

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -749,4 +753,4 @@ DEPENDENCIES
   webmock
 
 BUNDLED WITH
-   2.0.2
+   2.1.4
```
Comment (toyhammered) [inline Gemfile.lock:756]: @NuckChorris is it bad to just update the bundle version?

File: app/models/list_import/anilist.rb [hunk 2]
```diff
@@ -36,77 +12,122 @@ class Anilist < ListImport
     validate :ensure_user_exists, on: :create
 
     def ensure_user_exists
-      return if input_text.blank?
-
-      @auth_token ||= get_auth_token
```
Comment (matthewdias) [inline app/models/list_import/anilist.rb:41]: Don't we need to authenticate to get private entries?

File: app/models/list_import/anilist.rb [hunk 3]
```diff
@@ -36,77 +12,122 @@ class Anilist < ListImport
...
+      )
     end
+
+    MEDIA_LIST_QUERY = AnilistApiWrapper::Client.parse <<-'GRAPHQL'
+      query($user_name: String) {
+        anime: MediaListCollection(userName: $user_name, type: ANIME) {
+          lists {
+            name
+            entries {
+              score
```
Comment (matthewdias) [inline app/models/list_import/anilist.rb]: I think you need to specify `score(format: POINT_100)` to do calcs against it assuming a 100 point rating.

File: app/models/list_import/anilist/row.rb [hunk 4]
```diff
@@ -1,98 +1,126 @@
...
-      rescue
-        nil
+      def formatted_date(date_node)
+        date_node.to_h.symbolize_keys.values_at(:year, :month, :day).join('-').to_date.to_s
       end
 
-      def finished_at
-        return nil if node['finished_on'].nil?
+      def anilist_key
+        "AniList/#{type}"
```
Comment (matthewdias) [inline app/models/list_import/anilist/row.rb]: Can we do lower case `anilist/#{type}` to be consistent?

File: app/models/list_import/anilist/row.rb [hunk 5]
```diff
@@ -1,98 +1,126 @@
...
+      end
 
-        date = node['finished_on'].split('/')
+      def mal_key
+        "MyAnimeList #{type.capitalize}"
```
Comment (matthewdias) [inline app/models/list_import/anilist/row.rb]: Shouldn't this be `myanimelist/#{type}` ?

File: app/models/list_import/anilist/row.rb [hunk 6]
```diff
@@ -1,98 +1,126 @@
...
       end
 
-      def progress
-        type == 'anime' ? node['episodes_watched'] : node['chapters_read']
+      # reconsume_count -> repeat
+      def reconsume_count
+        node.repeat
       end
 
-      def volumes
```
Comment (matthewdias) [inline app/models/list_import/anilist/row.rb:42]: Shouldn't we still import this value? should be able to use `progress_volumes` right?

**Review by matthewdias — CHANGES_REQUESTED** [review]
Looking good! just pointed out a few things.

File: app/models/list_import/anilist.rb (hunk 2 above)
Comment (toyhammered) [reply app/models/list_import/anilist.rb:41]: We might. I think we require they make their list public though for us to import (we don't want to do any auth ideally)

File: app/models/list_import/anilist/row.rb (hunk 6 above)
Comment (toyhammered) [reply app/models/list_import/anilist/row.rb:42]: good catch! I totally forgot to readd this.

File: app/models/list_import/anilist/row.rb (hunk 5 above)
Comment (toyhammered) [reply app/models/list_import/anilist/row.rb]: Screen Shot 2020-04-24 at 8 07 43 AM
Rails admin is what I went by.

File: app/models/list_import/anilist/row.rb (hunk 4 above)
Comment (toyhammered) [reply app/models/list_import/anilist/row.rb]: sure

File: app/models/list_import/anilist.rb (hunk 2 above)
Comment (matthewdias) [reply app/models/list_import/anilist.rb:41]: If we're willing to tradeoff losing private individual entries then 👍 . We'll want to make sure we communicate that in the UI.

File: app/models/list_import/anilist/row.rb (hunk 5 above)
Comment (matthewdias) [reply app/models/list_import/anilist/row.rb]: Yeah I think that's just a name specified to make it human-readable in the admin panel. It doesn't actually tie back to anything. https://github.com/hummingbird-me/kitsu-server/blob/the-future/config/initializers/rails_admin.rb#L195

Comment (toyhammered) [reply app/models/list_import/anilist/row.rb]: ahhh....... fml... I hate you rails admin....

File: app/models/list_import/anilist/row.rb (hunk 6 above)
Comment (toyhammered) [reply app/models/list_import/anilist/row.rb:42]: I am not seeing a `progress_volume` anywhere. All I can see in the schema is `volumes_owned` which is updated from library entry (Volume)..... Well that makes no sense but I can shove the amount in there? 
Screen Shot 2020-04-24 at 5 32 15 PM

File: app/models/list_import/anilist.rb (hunk 2 above)
Comment (toyhammered) [reply app/models/list_import/anilist.rb:41]: @NuckChorris what is your take on this?

File: app/models/list_import/anilist/row.rb (hunk 6 above)
Comment (matthewdias) [reply app/models/list_import/anilist/row.rb:42]: `progress_volumes` is from anilist, and it should map to `volumes_owned` (misnomer) on our end

**Review by matthewdias — APPROVED** [review]
Approved pending an answer on the auth question from @NuckChorris

File: app/models/list_import/anilist.rb (hunk 2 above)
Comment (toyhammered) [reply app/models/list_import/anilist.rb:41]: Totally forgot to mention on the original one it used my personal keys so no way in hell did private ever work 🤷

Comment (NuckChorris) [reply app/models/list_import/anilist.rb:41]: Yeah all our imports work this way, it's something I'd like to solve in future but it's not a huge priority imo

File: app/models/list_import/anilist/row.rb [hunk 7]
```diff
@@ -1,98 +1,139 @@
...
+        other_mapping = Mapping.lookup(mal_key, mal_id) ||
+                        Mapping.guess(type.classify.safe_constantize, media_info)
+
+        if other_mapping.present?
+          Mapping.create(
+            item: other_mapping,
+            external_site: anilist_key,
+            external_id: media.id
+          )
+        end
```
Comment (NuckChorris) [inline app/models/list_import/anilist/row.rb:22]: Seems iffy to record the result of Mapping.guess imo

Could you make this only created for idMal mappings?

File: app/models/list_import/anilist.rb [hunk 8]
```diff
@@ -36,77 +12,122 @@ class Anilist < ListImport
...
-          # status[0] is always going to be either
-          # completed, on-hold, etc... just skip over that part
-          status[1].each do |media|
-            row = Row.new(media, type)
+        send("#{type}_list").each do |media|
+          row = Row.new(media, type)
 
-            yield row.media, row.data
-          end
+          yield row.media_mapping, row.data
```
Comment (NuckChorris) [inline app/models/list_import/anilist.rb]: Shouldn't this be row.media?

**Review by NuckChorris — CHANGES_REQUESTED** [review]
Few minor notes but otherwise looks great

File: app/models/list_import/anilist.rb (hunk 8 above)
Comment (toyhammered) [reply app/models/list_import/anilist.rb]: I actually changed to media_mapping thinking it was more understandable (and they have a media field that I reference by #media).

Comment (toyhammered) [reply app/models/list_import/anilist.rb]: I have changed back to be row.media though.

File: app/models/list_import/anilist/row.rb (hunk 7 above)
Comment (toyhammered) [reply app/models/list_import/anilist/row.rb:22]: done

**Review by NuckChorris — APPROVED** [review]

**Pull request merged**